src/group_theory/perm/fin.lean

```diff
@@ -62,3 +62,121 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+def cycle_range {n : ℕ} : fin n → perm (fin n)
+-- We could write this as a product over `list.range`,
+-- but the explicit `mk`s would become `coe`s and make the proofs a lot more annoying.
+-- We could also write this as two `if`s, but it is very annoying to prove that
+-- indeed results in a permutation.
+| ⟨0, _⟩ := 1
+| i@⟨nat.succ i', hi⟩ :=
+have i' < i'.succ := nat.lt_succ_self _,
+  swap ⟨0, lt_of_le_of_lt (nat.zero_le _) hi⟩ i * cycle_range ⟨i', this.trans hi⟩
+
+@[simp] lemma cycle_range_zero' {n : ℕ} (h : 0 < n) : cycle_range ⟨0, h⟩ = 1 := rfl
+@[simp] lemma cycle_range_zero {n : ℕ} : cycle_range (0 : fin n.succ) = 1 := rfl
+
+/-- The definition of `cycle_range i.succ`, where `i` is given as a natural number.
+
+This lemma is not really supposed to be used since it gives such an ugly right hand side,
+use `cycle_range_of_lt`, `cycle_range_self` and `cycle_range_of_gt` instead.
+-/
+lemma cycle_range_succ_aux {n : ℕ} {i : ℕ} (hi : i + 1 < n) :
+  cycle_range ⟨i + 1, hi⟩ = swap ⟨0, lt_of_le_of_lt (nat.zero_le _) hi⟩ ⟨i + 1, hi⟩ *
+    cycle_range ⟨i, (nat.lt_succ_self _).trans hi⟩ := rfl
+
+lemma cycle_range_of_gt {n : ℕ} {i j : fin n.succ} (h : i < j) :
+  cycle_range i j = j :=
+begin
+  obtain ⟨i, hi⟩ := i,
+  induction i with i ih,
+  { simp },
+  { have hi' := ((nat.lt_succ_self i).trans hi),
+    have h' := ((nat.lt_succ_self i).trans h),
+    rw [cycle_range_succ_aux, perm.mul_apply, ih hi' h', swap_apply_of_ne_of_ne],
+    { rintro rfl,
+      cases h' },
+    { rintro rfl,
+      have : i.succ ≠ i.succ := subtype.coe_injective.ne h.ne,
+      contradiction } },
+end
+
+-- We do `of_le` first to get a stronger induction hypothesis,
+-- then derive separate `of_lt` and `of_eq` results.
+-- TODO: is there a non-awkward way to state this for `fin n` instead of `fin n.succ`?
+lemma cycle_range_of_le {n : ℕ} {i j : fin n.succ} (h : j ≤ i) :
+  cycle_range i j = if j = i then 0 else j + 1 :=
+begin
+  obtain ⟨i, hi⟩ := i,
+  induction i with i ih,
+  { simp [le_antisymm h (nat.zero_le j)], },
+
+  rw [cycle_range_succ_aux, perm.mul_apply],
+  -- Handle the `of_eq` and `of_lt` cases separately.
+  split_ifs with heq,
+  { rw [heq, cycle_range_of_gt (subtype.mk_lt_mk.mpr (nat.lt_succ_self i)),
+        swap_apply_right, fin.mk_zero] },
+  have h := lt_of_le_of_ne h heq,
+  have j_lt_n : (j : ℕ) < n := lt_of_lt_of_le h (nat.lt_succ_iff.mp hi),
+  rcases lt_or_eq_of_le (nat.lt_succ_iff.mp h) with h | rfl,
+  { rw [ih _ h.le, if_neg, swap_apply_of_ne_of_ne],
+    { intro hj0,
+      have hj0 := congr_arg (coe : fin _ → ℕ) hj0,
+      rw [fin.coe_add_one j_lt_n] at hj0,
+      exact nat.succ_ne_zero _ hj0 },
+    { intro hji,
+      have hji := congr_arg (coe : fin _ → ℕ) hji,
+      rw [fin.coe_add_one j_lt_n] at hji,
+      have : (j + 1 : ℕ) ≠ i + 1 := (nat.succ_lt_succ h).ne,
+      contradiction },
+    { intro hji,
+      have hji := congr_arg (coe : fin _ → ℕ) hji,
+      exact h.ne hji } },
```

Comment:
I don't like how this part of the proof looks. It's repetitive, but not exactly identical. Anyone have a nice idea to golf this?